src/trusted/service_runtime/nacl_syscall_common.c

```diff
@@ -4330,3 +4320,104 @@ int32_t NaClSysSelect (struct NaClAppThread *natp, int nfds, fd_set * readfds,
+
+    if ((void*) kNaClBadAddress == sysact) {
+        NaClLog(2, "NaClSysSigaction could not translate act, returning %d\n", -NACL_ABI_EFAULT);
+        return -NACL_ABI_EFAULT;
+    }
+  }
+
+  if (oact) {
+    sysoact = (struct nacl_abi_sigaction*) NaClUserToSysAddrRangeProt(nap, (uintptr_t) oact, sizeof(struct nacl_abi_sigaction), NACL_ABI_PROT_WRITE);
+
+    if ((void*) kNaClBadAddress == sysoact) {
+        NaClLog(2, "NaClSysSigaction could not translate oact, returning %d\n", -NACL_ABI_EFAULT);
+        return -NACL_ABI_EFAULT;
+    }
+  }
+
+  ret = lind_sigaction(sig, sysact, sysoact, nap->cage_id);
+  return ret;
+}
+
+int32_t NaClSysKill(struct NaClAppThread *natp, int32_t targetcageid, int32_t sig) {
+  struct NaClApp *nap = natp->nap;
+  return lind_kill(targetcageid, sig, nap->cage_id);
+}
+
+int32_t NaClSysSigprocmask(struct NaClAppThread *natp, int32_t how, const uint64_t *nacl_set, uint64_t *nacl_oldset) {
+  int32_t ret;
+  struct NaClApp *nap = natp->nap;
+  const uint64_t *sysset = NULL;
+  uint64_t *sysoldset = NULL;
+
+  if (nacl_set) {
+    sysset = (const uint64_t *) NaClUserToSysAddrRangeProt(nap, (uintptr_t) nacl_set, sizeof(uint64_t), NACL_ABI_PROT_READ);
+
+    if ((void *) kNaClBadAddress == sysset) {
+        NaClLog(2, "NaClSysSigprocmask could not translate set, returning %d\n", -NACL_ABI_EFAULT);
+        return -NACL_ABI_EFAULT;
+    }
+  }
+
+  if (nacl_oldset) {
+    sysoldset = (uint64_t *) NaClUserToSysAddrRangeProt(nap, (uintptr_t) nacl_oldset, sizeof(uint64_t), NACL_ABI_PROT_READ);
+
+    if ((void *) kNaClBadAddress == sysoldset) {
+        NaClLog(2, "NaClSysSigprocmask could not translate oldset, returning %d\n", -NACL_ABI_EFAULT);
+        return -NACL_ABI_EFAULT;
+    }
+  }
+
+  ret = lind_sigprocmask(how, sysset, sysoldset, nap->cage_id);
+  return ret;
+}
+
+int32_t NaClSysLindsetitimer(struct NaClAppThread *natp, int32_t which, const struct itimerval *new_value, struct itimerval *old_value) {
+  int32_t ret;
+  struct NaClApp *nap = natp->nap;
+  const struct itimerval *sys_new_value = NULL;
+  struct itimerval *sys_old_value = NULL;
+
+  if (new_value) {
+    sys_new_value = (const struct itimerval *) NaClUserToSysAddrRangeProt(nap, (uintptr_t) new_value, sizeof(struct itimerval), NACL_ABI_PROT_READ);
+
+    if ((void *) kNaClBadAddress == sys_new_value) {
+      NaClLog(2, "NaclSysLindsetitimer could not translate new_value, returning %d\n", -NACL_ABI_EFAULT);
+      return -NACL_ABI_EFAULT;
+    }
+  }
+
+  if (old_value) {
+    sys_old_value = (struct itimerval *) NaClUserToSysAddrRangeProt(nap, (uintptr_t) old_value, sizeof(struct itimerval), NACL_ABI_PROT_WRITE);
+
+    if ((void *) kNaClBadAddress == sys_old_value) {
+      NaClLog(2, "NaclSysLindsetitimer could not translate old_value, returning %d\n", -NACL_ABI_EFAULT);
+      return -NACL_ABI_EFAULT;
+    }
+  }
```

Comment:
Similar to the previous function, there are repeated patterns in NaClSysLindsetitimer that can be refactored. For example:

```C
int32_t NaClSysLindsetitimer(struct NaClAppThread *natp, int32_t which, const struct itimerval *new_value, struct itimerval *old_value) {
    struct NaClApp *nap = natp->nap;
    const struct itimerval *sys_new_value = NaClTranslateAddressForRead(nap, new_value, "new_value", sizeof(struct itimerval));
    struct itimerval *sys_old_value = (struct itimerval *) NaClTranslateAddressForWrite(nap, old_value, "old_value", sizeof(struct itimerval));

    if (!sys_new_value || !sys_old_value) {
        return NACL_SETITIMER_FAILURE;
    }

    return lind_lindsetitimer(which, sys_new_value, sys_old_value, nap->cage_id);
}

const void *NaClTranslateAddressForRead(struct NaClApp *nap, const void *addr, const char *name, size_t size) {
    return NaClTranslateAddressWithProtection(nap, addr, name, size, NACL_ABI_PROT_READ);
}

void *NaClTranslateAddressForWrite(struct NaClApp *nap, void *addr, const char *name, size_t size) {
    return (void *) NaClTranslateAddressWithProtection(nap, addr, name, size, NACL_ABI_PROT_WRITE);
}

const void *NaClTranslateAddressWithProtection(struct NaClApp *nap, const void *addr, const char *name, size_t size, int protection) {
    const void *sys_addr = NaClUserToSysAddrRangeProt(nap, (uintptr_t) addr, size, protection);
    
    if (NACL_SETITIMER_BAD_ADDRESS == sys_addr) {
        NaClLog(2, "NaclSysLindsetitimer could not translate %s, returning %d\n", name, NACL_SETITIMER_FAILURE);
        return NULL;
    }

    return sys_addr;
}
```
1. Separated the translation of addresses for reading and writing into their own utility functions.
2. Created a more generic utility function NaClTranslateAddressWithProtection to handle the repeated pattern of translating addresses with a given protection.